### DISCUSSION TITLE
Can the gitoxide library do the equivalent of `git clone --mirror REPO`?

### DISCUSSION THREAD

**Discussion by bradlarsen** [GitoxideLabs/gitoxide#738] [Q&A] [answered]
Gitoxide is used for walking Git repository objects in [Nosey Parker](https://github.com/praetorian-inc/noseyparker), a fast detector of hardcoded secrets.

I'm adding functionality to Nosey Parker to be able to scan for secrets in a Git repository without first manually cloning it; instead, just specifying the URL will be enough. In this secrets detection use case, it's preferable to clone Git repositories with the `--mirror` option, as that may make additional objects accessible relative to a regular `git clone`.

The code in Nosey Parker is currently shelling out to the `git` binary, which is suboptimal in many ways. I'd rather do the mirrored cloning programmatically. In a previous implementation, I had tried to use `libgit2`, but its support for mirrored cloning was brittle and differed behavior-wise from the canonical `git` program.

Can the gitoxide library do the equivalent of `git clone --mirror REPO`? I couldn't figure this out from perusing API docs. I'd love to use gitoxide instead.

Thanks in advance!

**Comment by Byron** [comment, marked as answer]
At this time `mirror` isn't directly supported. After short investigation, it would mean the following to implement it.

1. fetch every ref there is on the remote
2. set the `remote.origin.mirror` key to true in the remote configuration to make pushes behave like a mirror (and probably push everything each time), and fetch all refs in all future fetches.

For clones one would be able to configure the ref-specs prior to the clone to capture all refs advertised by the remote, and for fetches one could do the same, even on the caller side. If you were inclined to implemented on your end, then I believe that would already work. I have also put it [into the roadmap](https://github.com/Byron/gitoxide/issues/470) *(search for `--mirror`)* for eventual implementation, but there is no timeline and I can't promise it's going to happen this year.

**Reply by bradlarsen** [reply]
Thanks @Byron !

**Comment by CodyCline** [comment]
@bradlarsen with a similar purpose in mind and based on @Byron answer, here is a barebones equivalent of `git clone --mirror url` 

This uses gix >= 0.73 for reference and should download all advertised objects, it seems that trying to specify `+refs/*:refs/*` will NOT work with gitoxide and a fetch request must be specific/verbose. As for what's missing you'll have to figure out modifying the config file values and measure performance impacts.  

```rs
use anyhow::{anyhow, Context, Result};
use bstr::BString;
use gix::{remote::fetch::Shallow, Repository, ThreadSafeRepository};
use std::{
    borrow::Cow,
    ffi::OsStr,
    path::{Path, PathBuf},
    sync::Arc,
};

pub fn mirror_clone(
    url: impl AsRef<OsStr>,
    directory: Option<impl Into<std::path::PathBuf>>,
    bare: bool,
    include_tags: bool,
) -> Result<(), anyhow::Error> {
    let url: gix::Url = url.as_ref().try_into()?;

    let directory = directory.map_or_else(
        || {
            let path = gix::path::from_bstr(Cow::Borrowed(url.path.as_ref()));
            path.file_name()
                .map(Into::into)
                .context("Filename extraction failed - path too short")
        },
        |dir| Ok(dir.into()),
    )?;
    let repo = gix::init_bare(directory)?;

    let mut remote = repo
        .remote_at(url)?
        .with_fetch_tags(gix::remote::fetch::Tags::None);

    let res = remote.connect(gix::remote::Direction::Fetch)?;

    //This is essentially an ls-remote equivalent
    let (ref_map, _outcome): (
        gix::remote::fetch::RefMap,
        gix::protocol::handshake::Outcome,
    ) = res.ref_map(
        gix::progress::Discard,
        gix::remote::ref_map::Options::default(),
    )?;

    //Map the refspecs out
    let rspecs: Vec<BString> = ref_map
        .remote_refs
        .iter()
        .map(|rs| rs.unpack().0.to_owned())
        .collect();
    remote = remote.with_refspecs(rspecs, gix::remote::Direction::Fetch)?;


    gix::remote::fetch::Outcome = remote
        .connect(gix::remote::Direction::Fetch)?
        .prepare_fetch(gix::progress::Discard, Default::default())?
        .with_shallow(Shallow::NoChange)
        .receive(gix::progress::Discard, &gix::interrupt::IS_INTERRUPTED)?;
    
    Ok(())
}
```

**Reply by Byron** [reply]
Thanks for sharing! 

`+refs/*:refs/*`  doesn't look like a valid refspec to me - is there a Git example somewhere that makes use of it? A code reference would certainly do it as well if you or somebody is inclined.

**Reply by CodyCline** [reply]
I must confess that I am not the most adept with all this git terminology. When you run `git clone --mirror https://github.com/GitoxideLabs/gitoxide.git` and then `cat gitoxide.git/config` it should look like this

```toml
[core]
	repositoryformatversion = 0
	filemode = true
	bare = true
[remote "origin"]
	url = https://github.com/GitoxideLabs/gitoxide
	fetch = +refs/*:refs/*
	mirror = true

```
Isn't that fetch parameter a refspec?

**Reply by Byron** [reply]
Ah, right, this is all I needed to see, thank you! This means behind a  mirror, the refspec is `+refs/*:refs/*`, which makes sense, and I think without tests it's clear that `gix` doesn't currently handle that correctly. I think it should, in principle, but I guess it's not really doing that in practice. Fair enough.
